Support marshaling from and unmarshaling to *types.Type

ehalpern opened the pull request:
Marshaling from `*types.Type` produces a noms type:
```
type S struct { Type *types.Type }
go := S{ types.String }
noms := types.NewStruct("S", types.StructData{"type": types.String})
v, err := marshal.Marshal(go) 
assert.Equals(t, noms, v)
```

Unmarshaling to `*types.Type` produces a pointer to an instance of the value:
```
type S struct { Type *types.Type }
noms := types.NewStruct("S", types.StructData{"type": types.String})
go :=  S{ types.String }

var s S
err := marshal.Unmarshal(v, &s) 
assert.Equals(t, go, s)
```

fixes: #2889

ehalpern commented:
cc: @willhite

arv commented:
In the PR description: types.String -> types.StringType

arv commented on go/marshal/decode.go:
```diff
@@ -144,6 +144,12 @@ func typeDecoder(t reflect.Type, tags nomsTags) decoderFunc {
+		if t.Implements(nomsValueInterface) {
+			return nomsValueDecoder
+		}
+		panic(&UnsupportedTypeError{Type: t})
```
you can also use a `fallthrough` statement

aboodman commented:
Documentation at top of encode.go and decode.go needs updating.

ehalpern commented:
Addressed review suggestions. 
@aboodman: Thx for the catch on godocs. I updated decode, but left encode as is since it already indicates that noms values are marshaled as is.
